README.md
PostfixAdmin Cookbook
=====================
[![GitHub](http://img.shields.io/badge/github-zuazo/postfixadmin--cookbook-blue.svg?style=flat)](https://github.com/zuazo/postfixadmin-cookbook)
[![License](https://img.shields.io/github/license/zuazo/postfixadmin-cookbook.svg?style=flat)](#license-and-author)

[![Cookbook Version](https://img.shields.io/cookbook/v/postfixadmin.svg?style=flat)](https://supermarket.chef.io/cookbooks/postfixadmin)
[![Dependency Status](http://img.shields.io/gemnasium/zuazo/postfixadmin-cookbook.svg?style=flat)](https://gemnasium.com/zuazo/postfixadmin-cookbook)
[![Code Climate](http://img.shields.io/codeclimate/github/zuazo/postfixadmin-cookbook.svg?style=flat)](https://codeclimate.com/github/zuazo/postfixadmin-cookbook)
[![Build Status](http://img.shields.io/travis/zuazo/postfixadmin-cookbook.svg?style=flat)](https://travis-ci.org/zuazo/postfixadmin-cookbook)
[![Coverage Status](http://img.shields.io/coveralls/zuazo/postfixadmin-cookbook.svg?style=flat)](https://coveralls.io/r/zuazo/postfixadmin-cookbook?branch=master)

Installs and configures [PostfixAdmin](http://postfixadmin.sourceforge.net/), a web based interface used to manage mailboxes, virtual domains and aliases.

Also creates the required *MySQL* or *PostgreSQL* database and tables.

Table of Contents
=================

- [Requirements](#requirements)
- [Generated Passwords](#generated-passwords)
- [Attributes](#attributes)
  - [The HTTPS Certificate](#the-https-certificate)
  - [Encrypted Attributes](#encrypted-attributes)
- [Recipes](#recipes)
  - [postfixadmin::default](#postfixadmindefault)
  - [postfixadmin::map_files](#postfixadminmap_files)
  - [postfixadmin::mysql](#postfixadminmysql)
  - [postfixadmin::postgresql](#postfixadminpostgresql)
- [Resources](#resources)
  - [postfixadmin_admin[user]](#postfixadmin_adminuser)
  - [postfixadmin_domain[domain]](#postfixadmin_domaindomain)
  - [postfixadmin_mailbox[mailbox]](#postfixadmin_mailboxmailbox)
  - [postfixadmin_alias[address]](#postfixadmin_aliasaddress)
  - [postfixadmin_alias_domain[address]](#postfixadmin_alias_domainaddress)
- [Usage Example](#usage-example)
  - [Including in a Cookbook Recipe](#including-in-a-cookbook-recipe)
  - [Including in the Run List](#including-in-the-run-list)
- [PostgreSQL Support](#postgresql-support)
  - [PostgreSQL Versions < 9.3](#postgresql-versions--93)
- [Deploy with Docker](#deploy-with-docker)
- [Testing](#testing)
  - [ChefSpec Matchers](#chefspec-matchers)
- [Contributing](#contributing)
- [TODO](#todo)
- [License and Author](#license-and-author)

Requirements
============

Please, if you want to upgrade the `postfixadmin` cookbook version from the `1.x` branch, see the [CHANGELOG](https://github.com/zuazo/postfixadmin-cookbook/blob/master/CHANGELOG.md#v200-2015-05-09).

## Supported Platforms

This cookbook has been tested on the following platforms:

* Amazon Linux
* CentOS
* Debian
* Fedora
* Ubuntu

Please, [let us know](https://github.com/zuazo/postfixadmin-cookbook/issues/new?title=I%20have%20used%20it%20successfully%20on%20...) if you use it successfully on any other platform.

## Required Cookbooks

* [apache2](https://supermarket.chef.io/cookbooks/apache2)
* [ark](https://supermarket.chef.io/cookbooks/ark)
* [database](https://supermarket.chef.io/cookbooks/database)
* [encrypted_attributes](https://supermarket.chef.io/cookbooks/encrypted_attributes)
* [mysql](https://supermarket.chef.io/cookbooks/mysql)
* [chef_nginx](https://supermarket.chef.io/cookbooks/chef_nginx)
* [compat_resource](https://supermarket.chef.io/cookbooks/compat_resource)
* [openssl](https://supermarket.chef.io/cookbooks/openssl)
* [php](https://supermarket.chef.io/cookbooks/php)
* [php-fpm](https://supermarket.chef.io/cookbooks/php-fpm)
* [postgresql](https://supermarket.chef.io/cookbooks/postgresql)
* [ssl_certificate](https://supermarket.chef.io/cookbooks/ssl_certificate)
* [yum-epel](https://supermarket.chef.io/cookbooks/yum-epel)

## Required Applications

* Chef `12.5` or higher.
* Ruby `2.2` or higher.

Only Postfix Admin version `3` or higher is supported by this cookbook. For older versions, use cookbook versions `< 3`.

## Other Requirements

On RedHat based platforms, you need to disable or configure SELinux correctly to work with `mysql` cookbook. You can use the `selinux::disabled` recipe for that.

Generated Passwords
===================

The first time it runs, automatically generates some passwords if not specified. Generated passwords are:

## From the PostfixAdmin Default Recipe

* `postfixadmin/setup_password`
* `postfixadmin/setup_password_salt`
* `postfixadmin/setup_password_encrypted`
* `postfixadmin/database/password`

## When MySQL Is Used

* `postfixadmin/mysql/server_root_password`

## When PostgreSQL Is Used

* `postgresql/password/postgres`

Attributes
==========

| Attribute                                             | Default                 | Description                    |
|:------------------------------------------------------|:------------------------|:-------------------------------|
| `node['postfixadmin']['version']`                     | `'3.0.2'`               | PostfixAdmin version
| `node['postfixadmin']['url']`                         | *calculated*            | PostfixAdmin download URL
| `node['postfixadmin']['checksum']`                    | *calculated*            | PostfixAdmin download file checksum
| `node['postfixadmin']['port']`                        | *calculated*            | PostfixAdmin listen port
| `node['postfixadmin']['server_name']`                 | *calculated*            | PostfixAdmin server name
| `node['postfixadmin']['server_aliases']`              | `[]`                    | PostfixAdmin server aliases
| `node['postfixadmin']['headers']`                     | `{}`                    | PostfixAdmin HTTP headers to set as hash
| `node['postfixadmin']['ssl']`                         | `false`                 | enables HTTPS (with SSL)
| `node['postfixadmin']['encrypt_attributes']`          | `false`                 | Whether to encrypt PostfixAdmin attributes containing credential secrets.
| `node['postfixadmin']['setup_password']`              | *calculated*            | PostfixAdmin Setup Password (required for chef-solo)
| `node['postfixadmin']['setup_password_salt']`         | *calculated*            | PostfixAdmin password salt (required for chef-solo)
| `node['postfixadmin']['web_server']`                  | `'apache'`              | Web server to use: `'apache'`, `'nginx'` or `false`
| `node['postfixadmin']['setup_password_encrypted']`    | *calculated*            | PostfixAdmin encrypted Password
| `node['postfixadmin']['database']['manage']`          | *calculated*            | Whether to manage database creation.
| `node['postfixadmin']['database']['type']`            | `'mysql'`               | PostfixAdmin database type. Possible values are: `'mysql'`, `'postgresql' (Please, see [below](#postgresql-support)<a></a>)`
| `node['postfixadmin']['database']['name']`            | `'postfix'`             | PostfixAdmin database name
| `node['postfixadmin']['database']['host']`            | `'127.0.0.1'`           | PostfixAdmin database hostname or IP address
| `node['postfixadmin']['database']['user']`            | `'postfix'`             | PostfixAdmin database login username
| `node['postfixadmin']['database']['password']`        | *calculated*            | PostfixAdmin database login password (requried for chef-solo)
| `node['postfixadmin']['mysql']['instance']`           | `'default'`             | PostfixAdmin MySQL instance name to run by the mysql_service LWRP from the mysql cookbook
| `node['postfixadmin']['mysql']['data_dir']`           | *calculated*            | PostfixAdmin MySQL data files path
| `node['postfixadmin']['mysql']['port']`               | `'3306'`                | PostfixAdmin MySQL port
| `node['postfixadmin']['mysql']['run_group']`          | *calculated*            | PostfixAdmin MySQL system group
| `node['postfixadmin']['mysql']['run_user']`           | *calculated*            | PostfixAdmin MySQL system user
| `node['postfixadmin']['mysql']['version']`            | *calculated*            | PostfixAdmin database MySQL version to install
| `node['postfixadmin']['conf']['encrypt']`             | `'md5crypt'`            | The way do you want the passwords to be crypted
| `node['postfixadmin']['conf']['domain_path']`         | `'YES'`                 | Whether you want to store the mailboxes per domain
| `node['postfixadmin']['conf']['domain_in_mailbox']`   | `'NO'`                  | Whether you want to have the domain in your mailbox
| `node['postfixadmin']['conf']['fetchmail']`           | `'NO'`                  | Whether you want fetchmail tab
| `node['postfixadmin']['packages']['requirements']`    | *calculated*            | PostfixAdmin required packages array
| `node['postfixadmin']['packages']['mysql']`           | *calculated*            | PostfixAdmin required packages array for MySQL support
| `node['postfixadmin']['packages']['postgresql']`      | *calculated*            | PostfixAdmin required packages array for PostgreSQL support
| `node['boxbilling']['mysql']['server_root_password']` | *calculated*            | PostfixAdmin MySQL *root* password.
| `node['postfixadmin']['map_files']['path']`           | `'/etc/postfix/tables'` | Path to generate map-files into
| `node['postfixadmin']['map_files']['mode']`           | `00640`                 | Map-files file-mode bits
| `node['postfixadmin']['map_files']['owner']`          | `'root'`                | Map-files files owner
| `node['postfixadmin']['map_files']['group']`          | `'postfix'`             | Map-files files group
| `node['postfixadmin']['map_files']['list']`           | *calculated*            | An array with map file names to generate
| `node['postfixadmin']['php-fpm']['pool']`             | `'postfixadmin'`        | PHP-FPM pool name to use with PostfixAdmin.

## The HTTPS Certificate

This cookbook uses the [`ssl_certificate`](https://supermarket.chef.io/cookbooks/ssl_certificate) cookbook to create the HTTPS certificate. The namespace used is `node['postfixadmin']`. For example:

```ruby
node.default['postfixadmin']['common_name'] = 'postfixadmin.example.com'
include_recipe 'postfixadmin'
```

See the [`ssl_certificate` namespace documentation](https://supermarket.chef.io/cookbooks/ssl_certificate#namespaces) for more information.

## Encrypted Attributes

This cookbook can use the [encrypted_attributes](https://supermarket.chef.io/cookbooks/encrypted_attributes) cookbook to encrypt the secrets generated during the *Chef Run*. This feature is disabled by default, but can be enabled setting the `node['postfixadmin']['encrypt_attributes']` attribute to `true`. For example:

```ruby
include_recipe 'encrypted_attributes::users_data_bag'
node.default['postfixadmin']['encrypt_attributes'] = true
inclure_recipe 'postfixadmin'
```

This will create the following encrypted attributes:

* `node['postfixadmin']['setup_password']`: PostfixAdmin *setup.php* setup password.
* `node['postfixadmin']['setup_password_encrypted']`: PostfixAdmin *setup.php* setup password encrypted with a salt.
* `node['postfixadmin']['mysql']['server_root_password']`: MySQL *root* user password.
* `node['postfixadmin']['database']['password']`: MySQL PostfixAdmin user password.

Read the [`chef-encrypted-attributes` gem documentation](http://onddo.github.io/chef-encrypted-attributes/) to learn how to read them.

**Warning:** When PostgreSQL is used, the database root password will still remain unencrypted in the `node['postgresql']['password']['postgres']` attribute due to limitations of the [postgresql cookbook](https://supermarket.chef.io/cookbooks/postgresql).

Recipes
=======

## postfixadmin::default

Installs and configures PostfixAdmin.

## postfixadmin::map_files

Installs PostfixAdmin SQL map files to be used by Postfix.

## postfixadmin::mysql

Installs MySQL server for PostfixAdmin.

## postfixadmin::postgresql

Installs PostgreSQL server for PostfixAdmin.

Resources
=========

## postfixadmin_admin[user]

Create or delete a PostfixAdmin admin user.

This kind of user is used to create the domains and mailboxes, and must be used before any other resource from this cookbook.

### postfixadmin_admin Actions

* `create`: Create a PostfixAdmin admin user (default).
* `delete`: Remove a PostfixAdmin admin user.

### postfixadmin_admin Properties

| Property       | Default                       | Description                    |
|:---------------|:------------------------------|:-------------------------------|
| user           | *name attribute*              | Username
| password       | *required*                    | Password
| setup_password | *calculated*                  | PostfixAdmin Setup Password
| superadmin     | `true`                        | Whether it has access to all domains
| domains        | `[]`                          | List of domains it has access to
| active         | `true`                        | Active status
| login_username | *optional*                    | Admin user to use for its creation
| login_password | *optional*                    | Admin password to use for its creation
| ssl            | `node['postfixadmin']['ssl']` | Whether to use SSL on HTTP requests

If you don't provide `login_username`, it will use the *setup.php* to create the admin. Usually this is used only to create the first administrator.

### postfixadmin_admin Example

```ruby
postfixadmin_admin 'admin@admindomain.com' do
  password 'sup3r-s3cr3t-p4ss'
  action :create
end

postfixadmin_admin 'secondadmin@admindomain.com' do
  password '4n0th3r-p4ss'
  login_username 'admin@admindomain.com'
  login_password 'sup3r-s3cr3t-p4ss'
end
```

## postfixadmin_domain[domain]

Create or delete a domain.

### postfixadmin_domain Actions

* `create`
* `delete`

### postfixadmin_domain Properties

| Property        | Default                       | Description                    |
|:----------------|:------------------------------|:-------------------------------|
| domain          | *name attribute*              | Domain name
| description     | `''`                          | Domain description
| aliases         | `10`                          | Maximum number of aliases
| mailboxes       | `10`                          | Maximum number of mailboxes
| active          | `true`                        | Active status
| default_aliases | `false`                       | Whether to include default aliases
| login_username  | *required*                    | Admin user to use
| login_password  | *required*                    | Admin password
| ssl             | `node['postfixadmin']['ssl']` | Whether to use SSL on HTTP requests

### postfixadmin_domain Example

```ruby
# admin user copied from the previous example
postfixadmin_domain 'foobar.com' do
  login_username 'admin@admindomain.com'
  login_password 'sup3r-s3cr3t-p4ss'
end
```

## postfixadmin_mailbox[mailbox]

Create or delete a mailbox.

### postfixadmin_mailbox Actions

* `create`
* `delete`

### postfixadmin_mailbox Properties

| Property       | Default                       | Description                    |
|:---------------|:------------------------------|:-------------------------------|
| mailbox        | *name attribute*              | Mailbox address to create
| password       | *required*                    | Mailbox password
| name           | `''`                          | The name of the mailbox owner
| active         | `true`                        | Active status
| mail           | `false`                       | Whether to send a welcome email
| login_username | *required*                    | Admin user to use
| login_password | *required*                    | Admin password
| ssl            | `node['postfixadmin']['ssl']` | Whether to use SSL on HTTP requests

### postfixadmin_mailbox Example

```ruby
# admin user copied from the previous example
postfixadmin_mailbox 'bob@foobar.com' do
  password 'alice'
  login_username 'admin@admindomain.com'
  login_password 'sup3r-s3cr3t-p4ss'
end
```

## postfixadmin_alias[address]

Create or delete a mailbox alias.

### postfixadmin_alias Actions

* `create`
* `delete`

### postfixadmin_alias Properties

| Property       | Default                       | Description                    |
|:---------------|:------------------------------|:-------------------------------|
| address        | *name attribute*              | Alias address
| goto           | *required*                    | Destination mailbox address
| active         | `true`                        | Active status
| login_username | *required*                    | Admin user to use
| login_password | *required*                    | Admin password
| ssl            | `node['postfixadmin']['ssl']` | Whether to use SSL on HTTP requests

### postfixadmin_alias Example

```ruby
# admin user copied from the previous example
postfixadmin_alias 'billing@foobar.com' do
  goto 'bob@foobar.com'
  login_username 'admin@admindomain.com'
  login_password 'sup3r-s3cr3t-p4ss'
end
```

## postfixadmin_alias_domain[address]

Create or remote a domain alias.

The domain name used as `alias_domain` must already exist: in other words, it needs to be created previously with `postfixadmin_domain` resource.

### postfixadmin_alias_domain Actions

* `create`

CHANGELOG.md
# Change Log
All notable changes to the `postfixadmin` cookbook will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Fixed
- Avoid clonning map files path directory resource with postfix-full.

## [3.0.0] - 2017-03-11
### Added
- Add `:active` and `:default_aliases` properties to domain resource.
- Add `:delete` action to resources.
- metadata: Add `chef_version`.
- Mark credentials as sensitive.

### Changed
- Update to install PostfixAdmin `3`.
- Use `chef_nginx` cookbook instead of the old `nginx`.
- Update some cookbook versions:
  - `mysql` from `6` to `8`.
  - `ark` from `0.9` to `2.2`.
  - `database` from `4` to `6`.
  - `php` from `1.5` to `2`.
  - `postgresql` from `3.4` to `6`.
  - `ssl_certificate` from `1.1` to `2`.
- Some LWRPs simplified: avoid making a DB connection.
- Use `Integer` instead of `Fixnum`.
- Fix all RuboCop and foodcritic offenses.
- Replace old LWRPs with Custom Resources.
- Update `openssl` cookbook password generation logic.
- Remove DB library completely, only use HTTP connections in the resources.
- Move libraries to `PostfixadminCookbook` namespace.
- CHANGELOG: Follow "Keep a CHANGELOG".

### Removed
- Drop Ruby `< 2.2` support.
- Drop Chef `< 12.5` support.
- Drop PostfixAdmin `2` installation support.
- Remove old `nginx` cookbook usage.
- Metadata: Remove grouping ([RFC-85](https://github.com/chef/chef-rfc/blob/8d47f1d0afa5a2313ed2010e0eda318edc28ba47/rfc085-remove-unused-metadata.md)).
- README: Remove documentation about locale (old).

### Fixed
- Always restart apache: improve Debian/Ubuntu support.
- Fix `postgresql_database_user[postfixadmin]` resource duplication.
- Fix Ubuntu `16.04` support.
- Fix `apache2` cookbook version `3.2` compatibility (issues [#9](https://github.com/zuazo/postfixadmin-cookbook/issues/9) and [#10](https://github.com/zuazo/postfixadmin-cookbook/issues/10), thanks [Mauro Destro](https://github.com/maurodx) and [Eric Blevins](https://github.com/e100) for reporting).
- Fix Chef `13` deprecation warnings.
- CHANGELOG: Fix capitals.

## [2.1.0] - 2015-08-21
### Added
- Add `node['postfixadmin']['database']['manage']` attribute.
- metadata: Add `source_url` and `issues_url`.
- Generate URL dynamically using the version value.
- Add Debian `8` and Ubuntu `15.04` support.

### Fixed
- Fix *session_start* error on CentOS.
- Fix `postfixadmin_admin#password`: it is not an encrypted attribute ([issue #6](https://github.com/zuazo/postfixadmin-cookbook/issues/6), thanks [Eric Blevins](https://github.com/e100) for reporting).

### Changed
- Update RuboCop to `0.33.0`.
- Update chef links to use *chef.io* domain.
- Update contact information and links after migration.
- README:
  - Use markdown tables.
  - Put the cookbook name in the title.

## [2.0.0] - 2015-05-09
### Changed
- Update the cookbook and the dependencies (**breaking change**):
- Update `mysql` cookbook to version `6`.
- Update `database` cookbook to version `4`.
- Update resources to use `mysql2` gem.
- Improve nginx support:
  - Restart nginx and php-fpm on first run for LWRP to work properly.
  - Fix php-fpm support on Ubuntu `10`.
- Update RuboCop to `0.30.1`.

### Upgrading from a `1.x.y` Cookbook Release
- **Note:*** Please do this with caution. Make a full backup before upgrading.

If you want to upgrade the cookbook [ersion from a `1.x` release, you should change the MySQL data directory path to the old one] - or migrate the database by hand:

```ruby
node.default['postfixadmin']['mysql']['data_dir'] = '/var/lib/mysql'
# [...]
include_recipe 'postfixadmin'
```

## [1.4.2] - 2015-05-07
### Fixed
- Mailbox resource: Fix name attribute to be a string.

## [1.4.1] - 2015-05-06
### Fixed
- Monkey patch the `MysqlClient#version` method missing error.

## [1.4.0] - 2015-02-13
- Fix disabling nginx default site.
- Remove `ssl_ca` param from `web_app` template.
- Include specific helpers instead of `"database::#{type}"` (issue [#5](https://github.com/zuazo/postfixadmin-cookbook/pull/5), thanks [Bernhard Weisshuhn (a.k.a. bernhorst)](https://github.com/bkw)).
  - Fixes `database` cookbook version `4` support.
- Run Unit tests against Chef `11` and `12`.
- Gemfile: Update RuboCop to `0.29.0`.

## [1.3.0] - 2015-01-05
- Add nginx support.
- Update `ssl_certificate` cookbook to `1.1.0`, adds chained certificate support.
- metadata: use pessimistic operator for cookbook versions, fixes database version bug.
- Bugfix: include OpenSSL functions in the `::postgresql` recipe.
- Berksfile: Use ark stable version for tests.
- Unit Tests: Update to use `ChefSpec::ServerRunner`.
- Fix Serverspec integration tests.
- travis.yml: Use the new build env.
- Gemfile:
  - Use foodcritic and RuboCop fixed versions.
  - Update to RuboCop `0.28.0`.
  - Update vagrant-wrapper to `2`.
- README:
  - Move the test matchers documentation to the README.
  - Add TOC.
  - *s/Attribute/Parameter/* for resources.
- TODO: add mysql cookbook update task.

## [1.2.0] - 2014-11-09
- ChefSpec matchers: added helper methods to locate LWRP resources.
- `PostfixAdmin::DB`: allow `db` instance attribute to be readable.
- Fix providers and map files password decryption with encrypted attributes enabled.
- Fix new RuboCop offense.
- Tests integrated with `should_not` gem.
- Update tests to work with Serverspec 2, includes Gemfile.
- Berksfile, Rakefile and Guarfile, generic templates copied.
- Enable ChefSpec coverage.
- TODO.md: Add some tasks.
- Homogenize license headers.

## [1.1.0] - 2014-10-03
- Added depends `mysql` cookbook `~> 5.0`.
- Include PHP recipe to fix Fedora support.
- Added LICENSE file.
- travis.yml: exclude some groups from bundle install.
- Gemfile:
  - Replaced `vagrant` git by `vagrant-wrapper`.
  - Berkshelf updated to `3.1`.
- Rakefile: require kitchen inside integration task.
- Guardfile added.

metadata.rb
# encoding: UTF-8
#
# Cookbook Name:: postfixadmin
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2013-2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

name 'postfixadmin'
maintainer 'Xabier de Zuazo'
maintainer_email 'xabier@zuazo.org'
license 'Apache 2.0'
description 'Installs and configures PostfixAdmin, a web based interface used '\
            'to manage mailboxes, virtual domains and aliases.'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version '3.1.0' # WiP

if respond_to?(:source_url)
  source_url "https://github.com/zuazo/#{name}-cookbook"
end
if respond_to?(:issues_url)
  issues_url "https://github.com/zuazo/#{name}-cookbook/issues"
end

chef_version '>= 12.5' if respond_to?(:chef_version)

supports 'amazon'
supports 'debian'
supports 'centos'
supports 'fedora'
supports 'ubuntu'

depends 'apache2', '~> 3.0'
depends 'ark', '~> 2.2'
depends 'database', '~> 6.0'
depends 'encrypted_attributes', '~> 0.2'
depends 'mysql', '~> 8.0'
depends 'chef_nginx', '~> 5.0'
depends 'compat_resource', '>= 12.5'
depends 'openssl', '~> 6.0'
depends 'php', '~> 2.0'
depends 'php-fpm', '>= 0.7'
depends 'postgresql', '~> 6.0'
depends 'mysql2_chef_gem', '~> 1.0'
depends 'ssl_certificate', '~> 2.0'
depends 'yum-epel', '~> 0.5'

recipe 'postfixadmin::default', 'Installs and configures PostfixAdmin'
recipe 'postfixadmin::map_files',
       'Installs PostfixAdmin SQL map files to be used by Postfix'
recipe 'postfixadmin::mysql',
       'Installs MySQL server for PostfixAdmin'
recipe 'postfixadmin::postgresql',
       'Installs PostgreSQL server for PostfixAdmin'

provides 'postfixadmin_admin'
provides 'postfixadmin_alias'
provides 'postfixadmin_domain'
provides 'postfixadmin_mailbox'
provides 'postfixadmin_alias_domain'
# Commented until 11.0.10 server release (CHEF-3976)
# provides 'postfixadmin_admin[user]'
# provides 'postfixadmin_alias[address]'
# provides 'postfixadmin_domain[domain]'
# provides 'postfixadmin_mailbox[mailbox]'
# provides 'postfixadmin_alias_domain[alias_domain]'

attribute 'postfixadmin/version',
          display_name: 'postfixadmin version',
          description: 'PostfixAdmin version',
          type: 'string',
          required: 'optional',
          default: '3.0.2'

attribute 'postfixadmin/url',
          display_name: 'postfixadmin URL',
          description: 'PostfixAdmin download URL',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'postfixadmin/checksum',
          display_name: 'postfixadmin checksum',
          description: 'PostfixAdmin download file checksum',
          type: 'string',
          required: 'optional',
          default:
            '761074e711ab618deda425dc013133b9d5968e0859bb883f10164061fd87006e'

attribute 'postfixadmin/port',
          display_name: 'postfixadmin port',
          description: 'PostfixAdmin listen port',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'postfixadmin/server_name',
          display_name: 'server name',
          description: 'PostfixAdmin server name',
          calculated: true,
          type: 'string',
          required: 'recommended'

attribute 'postfixadmin/server_aliases',
          display_name: 'server aliases',
          description: 'PostfixAdmin server aliases',
          type: 'array',
          required: 'optional',
          default: []

attribute 'postfixadmin/headers',
          display_name: 'postfixadmin headers',
          description: 'PostfixAdmin HTTP headers to set as hash',
          type: 'hash',
          default: {}

attribute 'postfixadmin/ssl',
          display_name: 'enable ssl',
          description: 'enables HTTPS (with SSL)',
          choice: %w(true false),
          type: 'string',
          required: 'optional',
          default: 'false'

attribute 'postfixadmin/encrypt_attributes',
          display_name: 'postfixadmin encrypt attributes',
          description: 'Whether to encrypt PostfixAdmin attributes containing '\
            'credential secrets.',
          type: 'string',
          choice: %w(true false),
          default: 'false'

attribute 'postfixadmin/setup_password',
          display_name: 'postfixadmin setup_password',
          description: 'PostfixAdmin Setup Password (required for chef-solo)',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'postfixadmin/setup_password_salt',
          display_name: 'postfixadmin password_salt',
          description: 'PostfixAdmin password salt (required for chef-solo)',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'postfixadmin/setup_password_encrypted',
          display_name: 'postfixadmin password_encrypted',
          description: 'PostfixAdmin encrypted Password',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'postfixadmin/web_server',
          display_name: 'Web Server',
          description: 'Web server to use: apache or false',
          choice: %w(apache nginx false),
          type: 'string',
          required: 'optional',
          default: 'apache'

attribute 'postfixadmin/database/manage',
          display_name: 'database manage',
          description: 'Whether to manage database creation.',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'postfixadmin/database/type',
          display_name: 'database type',
          description: 'PostfixAdmin database type',
          choice: %w("mysql" "postgresql"),
          type: 'string',
          required: 'optional',
          default: '"mysql"'

attribute 'postfixadmin/database/name',
          display_name: 'database name',
          description: 'PostfixAdmin database name',
          type: 'string',
          required: 'optional',
          default: '"postfix"'

attribute 'postfixadmin/database/host',
          display_name: 'database host',
          description: 'PostfixAdmin database hostname or IP address',
          type: 'string',
          required: 'optional',
          default: '"127.0.0.1"'

attribute 'postfixadmin/database/user',
          display_name: 'database user',
          description: 'PostfixAdmin database login username',
          type: 'string',
          required: 'optional',
          default: '"postfix"'

attribute 'postfixadmin/database/password',
          display_name: 'database password',
          description:
            'PostfixAdmin database login password (required for chef-solo)',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'postfixadmin/mysql/instance',
          display_name: 'mysql instance',
          description:
            'PostfixAdmin MySQL instance name to run by the mysql_service '\
            'LWRP from the mysql cookbook',
          type: 'string',
          required: 'optional',
          default: 'default'

attribute 'postfixadmin/mysql/data_dir',
          display_name: 'mysql data dir',
          description: 'PostfixAdmin MySQL data files path',
          type: 'string',
          required: 'optional',
          calcualted: true

attribute 'postfixadmin/mysql/port',
          display_name: 'mysql port',
          description: 'PostfixAdmin MySQL port',
          type: 'string',
          required: 'optional',
          default: '3306'

attribute 'postfixadmin/mysql/run_group',
          display_name: 'mysql run group',
          description: 'PostfixAdmin MySQL system group',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfixadmin/mysql/run_user',
          display_name: 'mysql run user',
          description: 'PostfixAdmin MySQL system user',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfixadmin/mysql/version',
          display_name: 'mysql version',
          description: 'PostfixAdmin MySQL version to install',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfixadmin/conf/encrypt',
          display_name: 'encryption configuration',
          description: 'The way do you want the passwords to be crypted',
          type: 'string',
          required: 'optional',
          default: '"md5crypt"'

attribute 'postfixadmin/conf/domain_path',
          display_name: 'domain path configuration',
          description: 'Whether you want to store the mailboxes per domain',
          choice: %w("YES" "NO"),
          type: 'string',
          required: 'optional',
          default: '"YES"'

attribute 'postfixadmin/conf/domain_in_mailbox',
          display_name: 'domain in mailbox configuration',
          description: 'Whether you want to have the domain in your mailbox',
          choice: %w("YES" "NO"),
          type: 'string',
          required: 'optional',
          default: '"NO"'

attribute 'postfixadmin/conf/fetchmail',
          display_name: 'enable fetchmail',
          description: 'Whether you want fetchmail tab',
          choice: %w("YES" "NO"),
          type: 'string',
          required: 'optional',
          default: '"NO"'

attribute 'postfixadmin/packages/requirements',
          display_name: 'postfixadmin packages requirements',
          description: 'PostfixAdmin required packages array',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfixadmin/packages/mysql',
          display_name: 'postfixadmin packages mysql',
          description: 'PostfixAdmin required packages array for MySQL support',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfixadmin/packages/postgresql',
          display_name: 'postfixadmin packages postgresql',
          description:
            'PostfixAdmin required packages array for PostgreSQL support',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'postfixadmin/mysql/server_root_password',
          display_name: 'postfixadmin mysql server root password',
          description: 'PostfixAdmin MySQL root password.',
          type: 'string',
          calculated: true

attribute 'postfixadmin/map_files/path',
          display_name: 'map files path',
          description: 'Path to generate map-files into',
          type: 'string',
          required: 'optional',
          default: '"/etc/postfix/tables"'

attribute 'postfixadmin/map_files/mode',
          display_name: 'map files mode',
          description: 'Map-files file-mode bits',
          type: 'string',
          required: 'optional',
          default: '00640'

attribute 'postfixadmin/map_files/owner',
          display_name: 'map files owner',
          description: 'Map-files files owner',
          type: 'string',
          required: 'optional',
          default: '"root"'

attribute 'postfixadmin/map_files/group',
          display_name: 'map files group',
          description: 'Map-files files group',
          type: 'string',
          required: 'optional',
          default: '"postfix"'

attribute 'postfixadmin/map_files/list',
          display_name: 'map files list',
          description: 'An array with map file names to generate',
          type: 'array',
          required: 'optional',
          default: [
            'db_virtual_alias_maps.cf',
            'db_virtual_alias_domain_maps.cf',
            'db_virtual_alias_domain_catchall_maps.cf',
            'db_virtual_domains_maps.cf',
            'db_virtual_mailbox_maps.cf',
            'db_virtual_alias_domain_mailbox_maps.cf',
            'db_virtual_mailbox_limit_maps.cf'
          ]

attribute 'postfixadmin/php-fpm/pool',
          display_name: 'postfixadmin php-fpm pool',
          description: 'PHP-FPM pool name to use with PostfixAdmin.',
          type: 'string',
          required: 'optional',
          default: 'postfixadmin'
